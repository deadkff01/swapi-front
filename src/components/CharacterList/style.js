import styled from 'styled-components';
import { Link } from '@reach/router';
import { Title } from 'components/Title';

export const ButtonLoadMore = styled.button`
  font-size: 1.2rem;
  font-weight: 600;
  text-transform: uppercase;
  background: transparent;
  border: 1px solid;
  border-radius: 5px;
  padding: 8px;
  transition: all 0.5s ease;

  &:hover {
    background: #fff;
    color: #000;
  }
`;

export const CharacterCardStyle = styled.div`
  font-weight: 600;
  &:hover {
    ${Title} {
      color: #d5d84b;
      text-decoration: underline;
    }
  }
`;

export function CharacterCard({ character }) {
  return (
    <CharacterCardStyle
      className="mt-5 md:w-full lg:w-1/3"
      data-testid="character-card"
    >
      <Link to={`/character/${character.url.split('/')[5]}`}>
        <Title className="text-2xl">{character.name}</Title>
      </Link>
    </CharacterCardStyle>
  );
}
