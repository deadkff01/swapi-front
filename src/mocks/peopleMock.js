export const peopleMock = [
  {
    name: 'Luke Skywalker',
    height: '172',
    mass: '77',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    birth_year: '19BBY',
    gender: 'male',
    homeworld: 'http://swapi.dev/api/planets/1/',
    films: [
      'http://swapi.dev/api/films/1/',
      'http://swapi.dev/api/films/2/',
      'http://swapi.dev/api/films/3/',
      'http://swapi.dev/api/films/6/'
    ],
    species: [],
    vehicles: [
      'http://swapi.dev/api/vehicles/14/',
      'http://swapi.dev/api/vehicles/30/'
    ],
    starships: [
      'http://swapi.dev/api/starships/12/',
      'http://swapi.dev/api/starships/22/'
    ],
    created: '2014-12-09T13:50:51.644000Z',
    edited: '2014-12-20T21:17:56.891000Z',
    url: 'http://swapi.dev/api/people/1/'
  },
  {
    name: 'C-3PO',
    height: '167',
    mass: '75',
    hair_color: 'n/a',
    skin_color: 'gold',
    eye_color: 'yellow',
    birth_year: '112BBY',
    gender: 'n/a',
    homeworld: 'http://swapi.dev/api/planets/1/',
    films: [
      'http://swapi.dev/api/films/1/',
      'http://swapi.dev/api/films/2/',
      'http://swapi.dev/api/films/3/',
      'http://swapi.dev/api/films/4/',
      'http://swapi.dev/api/films/5/',
      'http://swapi.dev/api/films/6/'
    ],
    species: ['http://swapi.dev/api/species/2/'],
    vehicles: [],
    starships: [],
    created: '2014-12-10T15:10:51.357000Z',
    edited: '2014-12-20T21:17:50.309000Z',
    url: 'http://swapi.dev/api/people/2/'
  },
  {
    name: 'R2-D2',
    height: '96',
    mass: '32',
    hair_color: 'n/a',
    skin_color: 'white, blue',
    eye_color: 'red',
    birth_year: '33BBY',
    gender: 'n/a',
    homeworld: 'http://swapi.dev/api/planets/8/',
    films: [
      'http://swapi.dev/api/films/1/',
      'http://swapi.dev/api/films/2/',
      'http://swapi.dev/api/films/3/',
      'http://swapi.dev/api/films/4/',
      'http://swapi.dev/api/films/5/',
      'http://swapi.dev/api/films/6/'
    ],
    species: ['http://swapi.dev/api/species/2/'],
    vehicles: [],
    starships: [],
    created: '2014-12-10T15:11:50.376000Z',
    edited: '2014-12-20T21:17:50.311000Z',
    url: 'http://swapi.dev/api/people/3/'
  }
];

export const C3PO = {
  name: 'C-3PO',
  height: '167',
  mass: '75',
  hair_color: 'n/a',
  skin_color: 'gold',
  eye_color: 'yellow',
  birth_year: '112BBY',
  gender: 'n/a',
  homeworld: 'http://swapi.dev/api/planets/1/',
  films: [
    'http://swapi.dev/api/films/1/',
    'http://swapi.dev/api/films/2/',
    'http://swapi.dev/api/films/3/',
    'http://swapi.dev/api/films/4/',
    'http://swapi.dev/api/films/5/',
    'http://swapi.dev/api/films/6/'
  ],
  species: ['http://swapi.dev/api/species/2/'],
  vehicles: [],
  starships: [],
  created: '2014-12-10T15:10:51.357000Z',
  edited: '2014-12-20T21:17:50.309000Z',
  url: 'http://swapi.dev/api/people/2/'
};
