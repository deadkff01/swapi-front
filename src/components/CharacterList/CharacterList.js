import { useEffect, useState, useCallback } from 'react';
import { useHttpInstance } from 'services/baseService';
import { Title } from 'components/Title';
import { ButtonLoadMore, CharacterCard } from './style';
import { ErrorWarning } from 'components/ErrorWarning';

export default function CharacterList({ characterList }) {
  const { post, response, error, loading } = useHttpInstance();
  const [peopleList, setPeopleList] = useState([]);
  const [lastIndex, setLastIndex] = useState(0);

  const getPeople = useCallback(async () => {
    const people = await post('people', {
      peopleArray: characterList,
      lastIndex: 0
    });
    if (response.ok) {
      setPeopleList(people);
      setLastIndex(getLastFromSliced(characterList, 0));
    }
  }, [post, response, characterList]);

  useEffect(() => {
    getPeople();
  }, [getPeople]);

  const isTheLast = () => characterList.length === lastIndex;

  const getLastFromSliced = (list, i) => {
    const sliced = list.slice(i, i + 3);
    const last = sliced[sliced.length - 1];
    return list.findIndex((l) => l === last) + 1;
  };

  const loadMorePeople = async () => {
    const body = {
      peopleArray: characterList,
      lastIndex
    };

    const getCurrentLast = getLastFromSliced(characterList, lastIndex);
    if (isTheLast()) {
      return;
    }
    setLastIndex(getCurrentLast);
    const people = await post('people', body);
    if (response.ok) {
      setPeopleList((p) => [...p, ...people]);
    }
  };

  if (error) {
    return <ErrorWarning message="Unable to load characters" />;
  }

  return (
    <div className="w-full mb-16" data-testid="characters-container">
      <Title className="px-5 text-3xl" align="left">
        Characters
      </Title>
      <div className="lg:flex flex-wrap justify-center mb-10">
        {peopleList.map((p, index) => (
          <CharacterCard key={index} character={p} />
        ))}
      </div>
      {loading ? (
        <Title className="text-md">Carregando personagens...</Title>
      ) : null}
      {!isTheLast() ? (
        <div className="flex justify-center">
          <ButtonLoadMore
            data-testid="load-more-button"
            onClick={loadMorePeople}
          >
            load more
          </ButtonLoadMore>
        </div>
      ) : null}
    </div>
  );
}
