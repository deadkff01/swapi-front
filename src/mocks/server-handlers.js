import { rest } from 'msw';
import { mockMovies, aNewHopeMock } from './mockMovies';
import { peopleMock, C3PO } from './peopleMock';

const handlers = [
  rest.get(`${process.env.REACT_APP_ENDPOINT}/movies`, (_, res, ctx) =>
    res(ctx.json(mockMovies))
  ),
  rest.get(`${process.env.REACT_APP_ENDPOINT}/movies/1`, (_, res, ctx) =>
    res(ctx.json(aNewHopeMock))
  ),
  rest.post(`${process.env.REACT_APP_ENDPOINT}/people`, (_, res, ctx) =>
    res(ctx.json(peopleMock))
  ),
  rest.get(`${process.env.REACT_APP_ENDPOINT}/people/2`, (_, res, ctx) =>
    res(ctx.json(C3PO))
  )
];

export { handlers };
