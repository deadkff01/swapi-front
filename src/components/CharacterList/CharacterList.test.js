import { cleanup, screen } from '@testing-library/react';
import { renderWithRouter } from '../../testUtils';
import '../../setupTests';
import CharacterList from './CharacterList';
import { aNewHopeMock } from 'mocks/mockMovies';
import { server } from 'mocks/server';

describe('CharacterList component', () => {
  beforeAll(() => server.listen());

  afterEach(() => server.resetHandlers());

  afterAll(() => {
    jest.useRealTimers();
    server.close();
    cleanup();
  });

  test('Renders the component', async () => {
    renderWithRouter(<CharacterList characterList={aNewHopeMock.characters} />);
    const container = await screen.findByTestId('characters-container');
    expect(container).toMatchSnapshot();
  });

  test('load character', async () => {
    renderWithRouter(<CharacterList characterList={aNewHopeMock.characters} />);
    const element = await screen.findAllByTestId('character-card');
    expect(element.length).toBe(3);
  });

  test('load message', async () => {
    renderWithRouter(<CharacterList characterList={aNewHopeMock.characters} />);
    const element = await screen.findByText('Carregando personagens...');
    expect(element).toBeDefined();
  });

  test('has load more button', async () => {
    renderWithRouter(<CharacterList characterList={aNewHopeMock.characters} />);
    const element = await screen.findByTestId('load-more-button');
    expect(element).toBeDefined();
  });
});
