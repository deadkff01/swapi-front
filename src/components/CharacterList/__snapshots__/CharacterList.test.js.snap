// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList component Renders the component 1`] = `
<div
  class="w-full mb-16"
  data-testid="characters-container"
>
  <h1
    class="sc-bdfBwQ dvefKP px-5 text-3xl"
  >
    Characters
  </h1>
  <div
    class="lg:flex flex-wrap justify-center mb-10"
  />
  <h1
    class="sc-bdfBwQ bZvnYH text-md"
  >
    Carregando personagens...
  </h1>
  <div
    class="flex justify-center"
  >
    <button
      class="sc-gsTCUz fLKFAS"
      data-testid="load-more-button"
    >
      load more
    </button>
  </div>
</div>
`;
